packages/runtime/src/runtime.ts

```diff
@@ -324,19 +347,160 @@ export class Runtime {
+   * @param freezeState target state
+   * @param payFlags transaction parameters
+   */
+  freezeAsset (
+    sender: string, assetId: number, freezeTarget: string,
+    freezeState: boolean, payFlags: TxParams
+  ): void {
+    this.mkAssetFreezeTx(sender, assetId, freezeTarget, freezeState, payFlags);
+    const acc = this.assertAccountDefined(freezeTarget, this.store.accounts.get(freezeTarget));
+    acc.freezeAsset(assetId, freezeState);
+  }
+
+  // creates asset revoke transaction object.
+  mkAssetRevokeTx (
+    sender: string, recipient: string, assetId: number,
+    revocationTarget: string, amount: number, flags: TxParams, assetID: number
+  ): void {
+    // this funtion is called only for validation of parameters passed
+    algosdk.makeAssetTransferTxnWithSuggestedParams(
+      sender,
+      recipient,
+      flags.closeRemainderTo,
+      revocationTarget,
+      amount,
+      encodeNote(flags.note, flags.noteb64),
+      assetId,
+      mockSuggestedParams(flags, this.round)
+    );
+    const asset = this.getAssetDef(assetID);
+    if (asset.clawback !== sender) {
+      throw new Error("Only Clawback account can revoke assets");
+    }
+  }
+
+  /**
+   * https://developer.algorand.org/docs/features/asa/#revoking-an-asset
+   * Revoking an asset for an account removes a specific number of the asset
+   * from the revoke target account.
+   * @param sender sender address
+   * @param recipient asset receiver address
+   * @param assetId asset index
+   * @param revocationTarget revoke target account
+   * @param amount amount of assets
+   * @param payFlags transaction parameters
+   */
+  revokeAsset (
+    sender: string, recipient: string,
+    assetID: number, revocationTarget: string,
+    amount: number, payFlags: TxParams
+  ): void {
+    this.mkAssetRevokeTx(sender, recipient, assetID, revocationTarget, amount, payFlags, assetID);
+    // Transfer assets
+    const fromAssetHolding = this.getAssetHolding(assetID, revocationTarget);
+    const toAssetHolding = this.getAssetHolding(assetID, recipient);
+    this.assertAssetNotFrozen(assetID, revocationTarget);
+    this.assertAssetNotFrozen(assetID, recipient);
+
+    if (fromAssetHolding.amount - amount < 0) {
+      throw new TealError(ERRORS.TRANSACTION.INSUFFICIENT_ACCOUNT_ASSETS, {
+        amount: amount,
+        address: revocationTarget
+      });
+    }
+    fromAssetHolding.amount -= amount;
+    toAssetHolding.amount += amount;
+  }
+
+  // creates asset destroy transaction object.
+  mkDestroyAssetTx (sender: string, assetId: number, flags: TxParams): void {
+    // this funtion is called only for validation of parameters passed
+    algosdk.makeAssetDestroyTxnWithSuggestedParams(
+      sender, encodeNote(flags.note, flags.noteb64),
+      assetId, mockSuggestedParams(flags, this.round)
+    );
+    const asset = this.getAssetDef(assetId);
+    if (asset.manager !== sender) {
+      throw new Error("Only Manager account can destroy assets");
```

Comment:
Please update this error as well. BDW - shouldn't this be a `TRANSACTION` / `ASA` error ?

Reply:
update to what ? yes, we can but we will only throw this error here, i don't think it will be repetitive.
If you think it will i can add all error in error list

Reply:
Update to `Only Manager account(<manager_address>) can destroy assets`, to make the error more intuitive. 
Don't have a particular preference regarding the error list. Your call.